samples/bluetooth/llpm/src/main.c

```diff
@@ -0,0 +1,415 @@
+	.error_found       = discovery_error,
+};
+
+static void advertise_and_scan(void)
+{
+	int err;
+
+	err = bt_le_adv_start(BT_LE_ADV_CONN, ad, ARRAY_SIZE(ad),
+			      sd, ARRAY_SIZE(sd));
+	if (err) {
+		printk("Advertising failed to start (err %d)\n", err);
+		return;
+	}
+
+	printk("Advertising successfully started\n");
+
+	err = bt_scan_start(BT_SCAN_TYPE_SCAN_PASSIVE);
+	if (err) {
+		printk("Starting scanning failed (err %d)\n", err);
+	}
+
+	printk("Scanning successfully started\n");
+}
+
+static void connected(struct bt_conn *conn, u8_t err)
+{
+	if (err) {
+		printk("Connection failed (err %u)\n", err);
+		return;
+	}
+
+	default_conn = bt_conn_ref(conn);
+	err = bt_conn_get_info(default_conn, &conn_info);
+	if (err) {
+		printk("Error %d while getting bt conn info\n", err);
+	}
+
+	printk("Connected as %s\n",
+	       conn_info.role == BT_CONN_ROLE_MASTER ? "master" : "slave");
+	printk("Conn. interval is %u units (1.25 ms/unit)\n",
+	       conn_info.le.interval);
+
+	/* make sure we're not scanning or advertising */
+	bt_le_adv_stop();
+	bt_scan_stop();
+
+	err = bt_gatt_dm_start(default_conn, BT_UUID_LATENCY, &discovery_cb,
+			       &gatt_latency);
+	if (err) {
+		printk("Discover failed (err %d)\n", err);
+	}
+}
+
+static void disconnected(struct bt_conn *conn, u8_t reason)
+{
+	printk("Disconnected (reason %u)\n", reason);
+
+	test_ready = false;
+
+	if (default_conn) {
+		bt_conn_unref(default_conn);
+		default_conn = NULL;
+	}
+
+	advertise_and_scan();
+}
+
+static void le_param_updated(struct bt_conn *conn, u16_t interval,
+			     u16_t latency, u16_t timeout)
+{
+	if (interval == INTERVAL_LLPM) {
+		printk("Connection interval updated: LLPM (1 ms)\n");
+	}
+}
+
+static int enable_llpm_mode(void)
+{
+	int err;
+	struct net_buf *buf;
+	hci_vs_cmd_llpm_mode_set_t *cmd_enable;
+
+	buf = bt_hci_cmd_create(HCI_VS_OPCODE_CMD_LLPM_MODE_SET,
+				sizeof(*cmd_enable));
+	if (!buf) {
+		printk("Could not allocate LLPM command buffer\n");
+	}
+
+	cmd_enable = net_buf_add(buf, sizeof(*cmd_enable));
+	cmd_enable->enable = true;
+
+	err = bt_hci_cmd_send(HCI_VS_OPCODE_CMD_LLPM_MODE_SET, buf);
+	if (err) {
+		printk("Error enabling LLPM %d\n", err);
+	} else {
+		printk("LLPM mode enabled\n");
+	}
+
+	return err;
+}
+
+static int enable_llpm_short_connection_interval(void)
+{
+	int err;
+	struct bt_le_conn_param conn_param = {
+		.interval_min = INTERVAL_LLPM,
+		.interval_max = INTERVAL_LLPM,
+		.latency = 0,
+		.timeout = 300
+	};
+
+	err = bt_conn_le_param_update(default_conn, &conn_param);
+	if (err) {
+		printk("Update connection parameters failed (err %d)\n", err);
+	}
+
+	return err;
+}
+
+static bool on_vs_evt(struct net_buf_simple *buf)
+{
+	u8_t code;
+	hci_vs_evt_qos_conn_event_report_t *evt;
+
+	code = net_buf_simple_pull_u8(buf);
+	if (code != HCI_VS_SUBEVENT_CODE_QOS_CONN_EVENT_REPORT) {
+		return false;
+	}
+
+	evt = (void *)buf->data;
+	if ((evt->event_counter & 0x3FF) == 0) {
+		printk("QoS conn event reports: "
```

Comment:
We only print the content of one report:
```suggestion
		printk("QoS conn event report: "
```
Maybe we should print the event counter as well?